Class/practice5.2/MegaDataPool.cpp

```diff
@@ -0,0 +1,57 @@
+#include "MegaDataPool.h"
+#include <stdexcept> // Include <stdexcept> for std::runtime_error
+
+MegaDataPool* MegaDataPool::instance = nullptr;
+std::mutex MegaDataPool::mutex; // Define static mutex variable
+
+MegaDataPool::MegaDataPool() {
+    // Private constructor
+}
+
+MegaDataPool& MegaDataPool::getInstance() {
+    // Ensure thread safety with double-checked locking
+    if (!instance) {
+        std::lock_guard<std::mutex> lock(mutex);
+        if (!instance) {
+            instance = new MegaDataPool();
+            instance->init(2); // Adjust the pool size as needed
+        }
+    }
+    return *instance;
+}
+
+void MegaDataPool::init(int poolSize) {
+    // Initialize the pool and availability vectors
+    pool.reserve(poolSize);
+    available.resize(poolSize, true);
+    for (int i = 0; i < poolSize; ++i) {
```

Comment:
what if pool already had some element then  we will get more elements then needed. I hope we should use not poolSize but something like std::max(0, poolSize - prevSize)